### ISSUE TITLE
can't write tile Code to transform NHWC to NCHW or transpose 1 matrix 

### ISSUE THREAD

**Issue by fsword73** [plaidml/plaidml#131]
# **For example,** 
func = plaidml.Function("""_**function (A[N,H,W,C]) -> (R[N,C,H,W]) {
    _R[n,c,h,w:N,C,H,W] = (A[n,h,w,c]);_
}**_""")

# **It reports error:** 
  File "C:\Python36\lib\site-packages\plaidml\__init__.py", line 853, in __init__
    code.encode(), fid.encode()))
  File "C:\Python36\lib\site-packages\plaidml\__init__.py", line 716, in _check_err
    self.raise_last_status()
  File "C:\Python36\lib\site-packages\plaidml\library.py", line 131, in raise_last_status
    raise self.last_status()
plaidml.exceptions.ResourceExhausted: Out of memory

# **The full code :** 

import numpy as np
import plaidml

#Generate input data
dimensions = (1,2, 2,3)
o_dimensions = (1,2, 2,3)

A = (np.random.rand(*dimensions) + 1.) / 52.
B = (np.random.rand(*dimensions) + 1.) / 52.
O = np.zeros(o_dimensions)
O_row_sum = O.sum(axis=1)
#Process data using PlaidML
ctx = plaidml.Context()

func = plaidml.Function("""function (A[N,H,W,C]) -> (R[N,C,H,W]) {
    R[n,c,h,w:N,C,H,W] = (A[n,h,w,c]);
}""")

with plaidml.open_first_device(ctx) as dev:

    dtype = plaidml.DType.FLOAT32
    in_shape = plaidml.Shape(ctx, dtype, *dimensions)
    a = plaidml.Tensor(dev, in_shape)
    with a.mmap_discard(ctx) as view:
        view[:] = A.flatten()
        view.writeback()
        
    b = plaidml.Tensor(dev, in_shape)
    with b.mmap_discard(ctx) as view:
        view[:] = B.flatten()
        view.writeback()
    
    out_shape = plaidml.Shape(ctx, dtype, *o_dimensions)    
    r = plaidml.Tensor(dev, out_shape)
    
    plaidml.run(ctx, func, inputs={"A": a}, outputs={"R": r})
    with r.mmap_current() as view:
        R = view[:]
print(R)    # Report results

**Comment by tzerrell** [comment]
There are 3 errors in the example code provided. I've included corrected code below but here is a summary of the problems:
1. The dimensions of the output data buffer are the same as the dimensions of the input data, but the Tile code reorders the dimensions. Correct this by changing `o_dimensions` to `(1, 3, 2, 2)`.
2. The dimensions of the output tensor are listed in the function header. The function header must not specify output dimensions; these are given only in the function body. Change `(R[N,C,H,W])` to `(R)` in the function header.
3. No [aggregation operation](https://vertexai-plaidml.readthedocs-hosted.com/en/latest/writing_tile_code.html#contractions) is given for the contraction. Even when a contraction is a simple assignment, Tile requires an aggregation operation to be explicitly specified. For assignment, `=` is the appropriate operation. Change `(A[n,h,w,c])` to `=(A[n,h,w,c])`.

Our apologies for the unclear "Out of memory" error message produced by this Tile syntax error. We are aware that our reporting of such errors is not user friendly and could use some attention.

In the example code below, I also removed unused code for clarity and added a print statement of the input data. This revised code runs correctly, at least on my machine (let us know if you have problems with it).

```
import numpy as np
import plaidml

#Generate input data
dimensions = (1,2, 2,3)
o_dimensions = (1,3, 2,2)

A = (np.random.rand(*dimensions) + 1.) / 52.
#Process data using PlaidML
ctx = plaidml.Context()

func = plaidml.Function("""function (A[N,H,W,C]) -> (R) {
R[n,c,h,w:N,C,H,W] = =(A[n,h,w,c]);
}""")

with plaidml.open_first_device(ctx) as dev:
    dtype = plaidml.DType.FLOAT32
    in_shape = plaidml.Shape(ctx, dtype, *dimensions)
    a = plaidml.Tensor(dev, in_shape)
    with a.mmap_discard(ctx) as view:
        view[:] = A.flatten()
        view.writeback()

    out_shape = plaidml.Shape(ctx, dtype, *o_dimensions)    
    r = plaidml.Tensor(dev, out_shape)

    plaidml.run(ctx, func, inputs={"A": a}, outputs={"R": r})
    with r.mmap_current() as view:
        R = view[:]
print(A)
print(R) # Report results
```

I also want to quickly mention that you are running PlaidML without a frontend, which adds a fair amount of complexity (e.g. this prints the output from `R` as an unshaped list since it doesn't include code managing the shape all the way to the end). If you are doing so to better understand the inner workings of PlaidML or to follow along in [our documentation](https://vertexai-plaidml.readthedocs-hosted.com/en/latest/life-function-api.html) this is reasonable.

However, if you want to use this code in an application I strongly encourage you to write this code as [an operation](https://vertexai-plaidml.readthedocs-hosted.com/en/latest/adding_ops.html). For example, here's how you would write this function as an Operation and run it with the Keras frontend:

```
import numpy as np
import plaidml
import plaidml.keras
plaidml.keras.install_backend()
import keras.backend as K


class NHWCtoNCHW(plaidml.tile.Operation):
    def __init__(self, x):
        code = """function (A[N,H,W,C]) -> (R) {
                      R[n,c,h,w:N,C,H,W] = =(A[n,h,w,c]);
                  }"""
        idims = x.shape.dims
        outshape = plaidml.tile.Shape(x.shape.dtype, (idims[0], idims[3], idims[1], idims[2]))
        super(NHWCtoNCHW, self).__init__(code, [('A', x)], [('R', outshape)])


#Generate input data
dimensions = (1, 2, 2, 3)
A = (np.random.rand(*dimensions) + 1.) / 52.
#Process data using PlaidML
input = K.variable(A)
result = NHWCtoNCHW.function(input)

print(A)
print(result.eval()) # Report results
```

**Comment by fsword73** [comment]
Cool!  Thanks a lot for your time.  I own you  a bottle of beer
